client/src/components/logViewer/index.ts

```diff
@@ -0,0 +1,56 @@
+// Copyright © 2024, SAS Institute Inc., Cary, NC, USA.  All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import {
+  DocumentSemanticTokensProvider,
+  OutputChannel,
+  SemanticTokensBuilder,
+  l10n,
+  window,
+} from "vscode";
+
+import { useStore } from "../../store";
+import { logSelectors } from "../../store/selectors";
+
+const { clearDataLogTokens } = useStore.getState();
+
+let outputChannel: OutputChannel;
+
+export const legend = {
+  tokenTypes: ["error", "warning", "note"],
+  tokenModifiers: [],
+};
+
+export const LogTokensProvider: DocumentSemanticTokensProvider = {
+  provideDocumentSemanticTokens: (document) => {
+    if (document.getText() === "") {
+      clearDataLogTokens();
+    }
+    const tokensBuilder = new SemanticTokensBuilder(legend);
+    const dataTokens = useStore.getState().logTokens;
+
+    for (let i = 0; i < dataTokens.length; i++) {
+      if (legend.tokenTypes.includes(dataTokens[i])) {
+        tokensBuilder.push(document.lineAt(i).range, dataTokens[i]);
+      }
+    }
+    return tokensBuilder.build();
+  },
+};
+
+useStore.subscribe(logSelectors.selectIsOutputChannelOpen, (open) => {
+  if (open) {
+    outputChannel.show(true);
+  }
+});
+
+useStore.subscribe(logSelectors.selectLogLines, (lines, prevLines) => {
+  if (!outputChannel) {
+    outputChannel = window.createOutputChannel(l10n.t("SAS Log"), "sas-log");
+  }
+
+  const delta = lines.filter((x) => !prevLines.includes(x));
```

Comment:
Lines here will be the entire new log lines state. We could try to use `OutputChannel.replace` here if we wanted to go that route (assuming we keep the output channel declaration in logViewer), but we would need to stringify the entire state before calling replace. On the first pass I had tried that, but it seemed inefficient to replace the entire contents of the OutputChannel for every log update.

Reply:
One motivating factor for capture log lines as state is that it would allow us down the road to offer a download/export log action which users have been asking for. VSCode has a way to programmatically access it (In digging it seems that the output channel is treated internally as a text document to vscode itself) but having a store action to dump those out seems preferable to having to go rooting around in vscode's document manager?

Reply:
Hmm... to me, storing the log lines doesn't necessarily mean storing as a state. For example a LogViewer can store the log lines internally and provide "append" and "get" methods.
Actually regarding download/export, getting from vscode's document sounds not a bad idea. SAS log may be huge and it can reduce some memory?